api/src/main/java/io/grpc/GlobalInterceptors.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2022 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+
+/** The collection of global interceptors and global server stream tracers. */
+@Internal
+public final class GlobalInterceptors {
+  private static GlobalInterceptors instance = null;
+  private List<ClientInterceptor> clientInterceptors;
+  private List<ServerInterceptor> serverInterceptors;
+  private List<ServerStreamTracer.Factory> serverStreamTracerFactories;
+  private boolean isGlobalInterceptorsTracersSet;
+  private boolean isGlobalInterceptorsTracersGet;
+
+  @VisibleForTesting
+  GlobalInterceptors() {}
+
+  /** Returns the GlobalInterceptors instance. */
+  private static synchronized GlobalInterceptors getInstance() {
+    if (instance == null) {
+      instance = new GlobalInterceptors();
+    }
+    return instance;
+  }
+
+  /**
+   * Sets the list of global interceptors and global server stream tracers.
+   *
+   * <p>If {@code setInterceptorsTracers()} is called again, this method will throw {@link
+   * IllegalStateException}.
+   *
+   * <p>It is only safe to call early. This method throws {@link IllegalStateException} after any of
+   * the get calls [{@link #getClientInterceptors()}, {@link #getServerInterceptors()} or {@link
+   * #getServerStreamTracerFactories()}] has been called, in order to limit changes to the result of
+   * {@code setInterceptorsTracers()}.
+   *
+   * @param clientInterceptorList list of {@link ClientInterceptor} that make up global Client
+   *     Interceptors.
+   * @param serverInterceptorList list of {@link ServerInterceptor} that make up global Server
+   *     Interceptors.
+   * @param serverStreamTracerFactoryList list of {@link ServerStreamTracer.Factory} that make up
+   *     global ServerStreamTracer factories.
+   */
+  public static void setInterceptorsTracers(
+      List<ClientInterceptor> clientInterceptorList,
+      List<ServerInterceptor> serverInterceptorList,
+      List<ServerStreamTracer.Factory> serverStreamTracerFactoryList) {
+    try {
+      GlobalInterceptors.getInstance()
+          .setGlobalInterceptorsTracers(
```

Comment:
Thinking about the synchronization, in this case `getInstance()` synchronizes on the class object and the non-static `setGlobalInterceptorsTracers` synchronizes on the instance variable.

Overall we want synchronization only for the callers of this (static `setInterceptorsTracers`) method and keep this structure for ease of unit testing. Would it make more sense to make this method `synchronized` and remove synchronization for both `getInstance` and `setGlobalInterceptorsTracers` ?